src/factorizations/blocklanczos.jl

```diff
@@ -8,57 +8,56 @@ while `S` represents the type of inner products between vectors. To create an in
 one can specify `S` explicitly and use `Block{S}(vec)`, or use `Block(vec)` directly, in which case an
 inner product is performed to infer `S`.
 """
-struct Block{T,S<:Number}
+struct Block{T}
     vec::Vector{T}
-    function Block{S}(vec::Vector{T}) where {T,S<:Number}
-        @assert length(vec) > 0
-        return new{T,S}(vec)
+    function Block{T}(vec::Vector{T}) where {T}
+        length(vec) > 0 || throw(ArgumentError("blocklength must be >(0)"))
+        return new{T}(vec)
     end
 end
-function Block(vec::Vector{T}) where {T}
-    @assert length(vec) > 0
-    S = typeof(inner(vec[1], vec[1]))
-    return Block{S}(vec)
-end # A convenient constructor for users
+Block(vec::Vector{T}) where {T} = Block{T}(vec)
+
 Base.length(b::Block) = length(b.vec)
-Base.getindex(b::Block, i::Int) = b.vec[i]
-function Base.getindex(b::Block{T,S}, idxs::AbstractVector{Int}) where {T,S}
-    return Block{S}([b.vec[i] for i in idxs])
-end
-Base.setindex!(b::Block{T}, v::T, i::Int) where {T} = (b.vec[i] = v)
-function Base.setindex!(b₁::Block{T}, b₂::Block{T},
-                        idxs::AbstractVector{Int}) where {T}
-    return (b₁.vec[idxs]=b₂.vec;
-            b₁)
+
+@inline Base.getindex(b::Block, i::Int) = b.vec[i]
+@inline Base.getindex(b::Block, idxs::AbstractVector{<:Integer}) = Block(b.vec[idxs])
+
+@inline Base.setindex!(b::Block{T}, v::T, i::Int) where {T} = (b.vec[i]=v; b)
+@inline function Base.setindex!(b₁::Block{T}, b₂::Block{T},
+                                idxs::AbstractVector{Int}) where {T}
+    b₁.vec[idxs] = b₂.vec
+    return b₁
 end
+
 LinearAlgebra.norm(b::Block) = norm(b.vec)
-function apply(f, block::Block{T,S}) where {T,S}
-    return Block{S}([apply(f, block[i]) for i in 1:length(block)])
-end
-function block_inner(B₁::Block{T,S}, B₂::Block{T,S}) where {T,S}
-    M = Matrix{S}(undef, length(B₁.vec), length(B₂.vec))
-    @inbounds for j in 1:length(B₂)
-        yj = B₂[j]
-        for i in 1:length(B₁)
-            M[i, j] = inner(B₁[i], yj)
-        end
+
+apply(f, block::Block) = Block(map(Base.Fix1(apply, f), block.vec))
+
+function VectorInterface.inner(B₁::Block{T}, B₂::Block{T}) where {T}
```

Comment:
Since `Block` is really a special type that is not supposed to represent a single vector, I would prefer to not make this another method definition of `VectorInterface.inner` (which, I would argue, returns a scalar value as part of the definition), and thus keep the `block_inner` name. There are some other `block_...` methods as well. I just noticed there is also a `blocklanczosrecurrence` which should have been `block_lanczosrecurrence` for consistency.